proposals/0000-surface-async-stream-errors-to-http.md

```diff
@@ -0,0 +1,231 @@
+
+```marko
+<!doctype html>
+<html lang="en">
+<head>
+  <PageAssets />
+  <PageMetadata />
+</head>
+
+<body>
+  <SiteHeader />
+
+  <PageContent>
+    <await(input.contentFetchRequest)>
+      <@then|response|>
+        ${response.body}
+      </@then>
+      <@catch|err|>
+        Oh no, the content API is down again
+      </@catch>
+    </await>
+  </PageContent>
+
+  <SiteFooter />
+</body>
+</html>
+```
+
+This is great for the user because their browser gets a head start downloading assets linked in the `<head>` and displaying the site header while the network request to the content API is underway.
+
+However, if the `contentFetchRequest` fails for any of the myriad reasons computers are terrible, the page will show an error message instead of its real content — but as far as the HTTP layer knows, the response completed successfully.
+
+No machine-readable indicator that the streamed response contains erroneous content causes some thorny problems:
+
+- An HTTP cache may store the erroneous content and reuse it, causing users to see the error for longer than they otherwise would
+
+- A search engine will index the erroneous content, since they received no sign they should try again or discard the response as invalid
+
+- HTTP-level tools (debuggers, curl, spiders, etc.) will report the response as successful, even when it wasn’t
+
+Research into HTTP/1.1’s chunked `Transfer-Encoding` and HTTP/2’s stream error handling found that they both have a standardized way of indicating a dynamically-streamed response failed to complete successfully:
+
+<dl>
+  <dt>HTTP/1.1 <code>Transfer-Encoding: chunked</code></dt>
+  <dd><a href="https://httpwg.org/http-core/draft-ietf-httpbis-messaging-latest.html#incomplete.messages">IETF Draft: HTTP/1.1 Messaging §8 Handling Incomplete Messages</a></dd>
+  <dd><p>If a chunked response doesn’t terminate with the zero-length end chunk, the client must assume that the response was <i>incomplete</i> — which at the very least, <a href="https://httpwg.org/http-core/draft-ietf-httpbis-cache-latest.html#rfc.section.3.1">means a cache should double-check with the server before reusing the stored incomplete response</a>.</p></dd>
+
+  <dt>HTTP/2 Streams</dt>
+  <dd><a href="https://tools.ietf.org/html/rfc7540#section-5.4.2">RFC 7540 §5.4.2 Stream Error Handling</a></dd>
+  <dd><p>An HTTP/2 stream can signal an application error by sending a <code>RST_STREAM</code> frame with an error code of <code>0x2 INTERNAL_ERROR</code>.</p></dd>
+</dl>
+
+This proposal explores how Marko should surface errors from server components to the HTTP layer, so the response can properly signal an error with one of the above methods.
+
+## Guide-level explanation
+
+Developers should be able to tell Marko how important an async error is:
+
+1. Our example from the Motivation section is as serious as it gets — you might even want to redirect to a completely different error page 
```

Comment:
Would it make sense to communicate show-stopping errors by not catching them?  We could just rely on the current uncaught functionality: the error is emitted on the stream and the app can choose to do something like write out a redirect script.
```html
<script>window.location = "https://www.ebay.com/n/error";</script>
```

Reply:
Another possible option would be to put the redirect script in the `<@catch>` block

```marko
<@catch>
   <script>window.location = "https://www.ebay.com/n/error";</script>
</@catch>
```

Reply:
Maybe an explicit function to call (`streamErrorRedirect()`, just as a strawman) would be better, if only because boy are there a lot of ways for JS to navigate:

- `window.location =`
- `window.location.href =`
- `location.assign()`
- `location.replace()`
- …and then multiply all of those by how `location` is also a global variable you can use without `window.`, _or_ you could access it off `document`

An explicit call could also let the developer redirect to various kinds of `5XX` errors, such as one that explicitly blames an API server the Marko front-end is gatewaying with (`502`, probably).

Reply:
As an API proposal, maybe something like:

```marko
<@catch|err, redirect|>
  <if(err === "Really Bad")>
    $ redirect("http://somepage.com")
  </if>
</@catch>
```

Reply:
Thinking about it, one problem with redirecting is that it prevents the user from hitting the Refresh button to see if that fixes the page, in the case of sporadic errors.

Reply:
I guess one work around would to be store the original page url and do some magic if the referrer is the error page itself to navigate back. Seems a bit strange though.

Reply:
I replied via email so GitHub didn’t know to put [my direct response to this](#issuecomment-612682751) here, so I’m quoting it:

> Yeah, especially since real HTTP errors don't change the URL address. Maybe redirecting isn't a good pattern in the first place; it's certainly unexpected for users, and there isn't a valid, clean way to express it in the bedrock HTML or HTTP layers.